auth.go

```diff
@@ -0,0 +1,192 @@
+package pgsrv
+
+import (
+	"bytes"
+	"crypto/md5"
+	"crypto/rand"
+	"fmt"
+)
+
+const expectedPasswordMessage = "expected password response, got message type %c"
+const passwordDidNotMatch = "Password does not match for user \"%s\""
+
+// authenticator interface defines objects able to perform user authentication
+// that happens at the very beginning of every session.
+type authenticator interface {
+	// authenticate accepts a msgReadWriter instance and a map of args that describe
+	// the current session. It returns true if the authentication succeeded,
+	// and the client was notified using the provided msgReadWriter object. It returns
+	// an error if something went wrong, even if the authentication was successful:
+	// e.g if it couldn't write a response into rw.
+	//
+	// Authentication errors as well as welcome messages are sent by this function,
+	// so there is no need for the caller to send these. It is caller's responsibility
+	// though to terminate the session in case that the first return value is false
+	// OR if the second returned value is not nil.
+	authenticate(rw msgReadWriter, args map[string]interface{}) (bool, error)
```

Comment:
@orlangure - why do we need to return two values here?
why not return const fmt.Errorf("authentication failed") here instead of false? is there any value of two cases that indicate an error?

Reply:
@aviaoh my point is that a failed auth is not the same as an error that occured somewhere in the process. Think of it as that if the second value exists (`err != nil`), then there is no definitive decision whether or not the user can access the system. If the second value is `nil` (no error), we know exactly whether or not to let the user in.

Reply:
but we are not letting the user in on both cases... 
authentication can failed with any error, either connection error or wrong password.

what is the difference from caller point of view?